plugin/pkg/auth/authorizer/node/node_authorizer.go

```diff
@@ -0,0 +1,169 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	"github.com/golang/glog"
+
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/kubernetes/pkg/api"
+	rbacapi "k8s.io/kubernetes/pkg/apis/rbac"
+	"k8s.io/kubernetes/pkg/auth/nodeidentifier"
+	"k8s.io/kubernetes/plugin/pkg/auth/authorizer/rbac"
+	"k8s.io/kubernetes/plugin/pkg/auth/authorizer/rbac/bootstrappolicy"
+	"k8s.io/kubernetes/third_party/forked/gonum/graph"
+	"k8s.io/kubernetes/third_party/forked/gonum/graph/traverse"
+)
+
+// NodeAuthorizer authorizes requests from kubelets, with the following logic:
+// 1. If a request is not from a node (IdentifyNode() returns isNode=false), reject
+// 2. If a request is not allowed by the rules in the default system:node cluster rule, reject
+// 3. If a specific node cannot be identified (IdentifyNode() returns nodeName=""):
+//    If in compatibility-mode (default), allow. This lets nodes that don't use node-specific identities continue to work with the broad authorization rules in step 2.
+//    If in strict-mode, reject. This lets deployments that provision all nodes with individual identities to indicate that only identifiable nodes should be allowed.
+// 4. If a request is for a secret, configmap, persistent volume or persistent volume claim, reject unless the verb is get, and the requested object is related to the requesting node:
+//    node -> pod
+//    node -> pod -> secret
+//    node -> pod -> configmap
+//    node -> pod -> pvc
+//    node -> pod -> pvc -> pv
+//    node -> pod -> pvc -> pv -> secret
+type NodeAuthorizer struct {
+	graph      *Graph
+	identifier nodeidentifier.NodeIdentifier
+	strict     bool
+	nodeRules  []rbacapi.PolicyRule
+}
+
+// New returns a new node authorizer
+func New(graph *Graph, identifier nodeidentifier.NodeIdentifier, strict bool) authorizer.Authorizer {
+	return &NodeAuthorizer{
+		graph:      graph,
+		identifier: identifier,
+		strict:     strict,
+		nodeRules:  bootstrappolicy.NodeRules(),
+	}
+}
+
+func (r *NodeAuthorizer) Authorize(attrs authorizer.Attributes) (bool, string, error) {
+	nodeName, isNode := r.identifier.NodeIdentity(attrs.GetUser())
+	if !isNode {
+		// reject requests from non-nodes
+		return false, "", nil
+	}
+	if len(nodeName) == 0 && r.strict {
+		// reject requests from unidentifiable nodes if in strict mode
+		return false, "unknown node", nil
+	}
+
+	if !rbac.RulesAllow(attrs, r.nodeRules...) {
+		// reject requests that don't fit within the broad authorization rules
+		return false, "", nil
+	}
+
+	if len(nodeName) == 0 {
+		// allow requests from unidentifiable nodes if not in strict mode
+		return true, "", nil
+	}
+
+	// subdivide access to specific resources
+
+	if attrs.GetAPIGroup() != api.GroupName {
```

Comment:
The list of `GroupResources` that you're checking is a clear whitelist.  I'd prefer a tight existence check. Since you'll want zero allocations I'm fine with an exploded if/else structure.